ons-spark/spark-app/spark_application_and_ui.md

```diff
@@ -0,0 +1,429 @@
+
+The first operation to be applied to this DataFrame is a *narrow* transformation, e.g. summing two columns to create a new column. The second operation is again a narrow transformation. These two operations can be run in parallel, because each child partition has only one parent, indicated by the golden arrows.
+
+Then we hit a stage boundary, which means the third operation is a *wide* transformation, e.g. aggregating the data or joining with another DataFrame. This is a wide operation because each child partition has multiple parent partitions. This is also an example of *shuffling* the data, which means to move data between partitions, more on shuffling in the **Shuffling Article**.
+
+Finally in the diagram above there is another narrow operation on the aggregated DataFrame before we run into an *action*, e.g. writing the data to HDFS or a simple count of the rows. 
+
+As we execute the wide and narrow transformations nothing happens to the data at this point in time, Spark just builds up a plan of what to do with the DataFrame. This plan is a called the execution plan and is a set of instructions of how to transform the data from one state to another. More on the execution plan in the **Persisting Article**. An action initiates the execution plan and so this is when the DataFrame is processed. All the previous transformations, along with the action, are put into Spark *jobs* and deployed on the Spark cluster. 
+
+## Investigate the Spark UI
+
+We will create an application and execute some code to create jobs. Then we can drill down from Job to Stage to Task and investigate their performance. In most cases this is where we look first when trying to diagnose a slow Spark application.
+
+### Create an application
+
+The first step is to create an application. Once we create a Spark application we can look at the Spark UI. If we stop the session using `spark.stop()`/`spark_disconnect()` or stopping the CDSW session, the Spark UI for our application can no longer be accessed.
+
+Let's start with the necesarry imports and create an application.
+````{tabs}
+```{code-tab} py
+import os, IPython
+import yaml
+from pyspark.sql import SparkSession, functions as F
+
+spark = SparkSession.builder.master("local[2]").appName("spark-app-ui").getOrCreate()
+```
+
+```{code-tab} r R
+
+library(sparklyr)
+
+default_config <- sparklyr::spark_config()
+
+sc <- sparklyr::spark_connect(
+    master = "local[2]",
+    app_name = "spark-app-ui",
+    config = default_config)
+
+```
+````
+
+```plaintext
+```
+The cell below creates a handy button for us to access the Spark UI. Make sure you hold down Ctrl when you click on the link to open the page in a new tab. See the [Cloudera documentation](https://docs.cloudera.com/documentation/data-science-workbench/1-0-x/topics/cdsw_spark_ui.html) for more informaiton on this code snippet.
```

Comment:
Typo: information.

I've got for the approach of just printing the URL out, rather than a link, as the output here is <IPython.core.display.HTML object>

Reply:
happy to use your method. Do you think there could be a security risk involved with having the domain visible in the book? There shouldn't be any data on dev anyway, and it's only accessible on the network, but thought I had better ask the question.

Reply:
Let's go with it for now and I'll ask Nigel at our catch up on Friday

Reply:
Looking at this again, now you're using the local host I think you can just reference it in the text rather than printing it out.

It was sensible to print it out before as the URL changed every time. Now it doesn't I don't think there is any point in printing it.